### PR TITLE
KeyCode Encryption

### PR DESCRIPTION
**Pull request by Pingu1337**

### CODE REVIEW

**Review by telamon — CHANGES_REQUESTED** [review]
General review Thumbs up, the initial feature is complete with minor fixes;

humble request, for the sake of project consistency try to use UpperCamel only for class names.

Good news, we have full TDD-support; Changes to `blockend/` can/should be developed using tests that don't require you to use the browser at all `yarn test`
and  `yarn debug`.
It'll be our only lifeline to sanity when network interactions gets thrown into the mix. 🤪

Please remove `package-json.lock` it's a very bad file :(
But to be fair I'm leaning towards discarding `yarn.lock` too..
(Something wierd happened to yarn's lockfile, did npm rewrite it? O_o)

##  This... Is... Brutally Awesome!
image


Thank you.

File: frontend/components/keypad-dialog.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    } else if (e.target.accessKey === 'BackspaceBtnUnlock') {
+      const NewVal = RemoveLastChar(KeyPadUnlockVal)
+      KeyPadUnlock.value = NewVal
+    }
+  }
+
+  render () {
+    if (this.props.post === 'false') {
+      return this.html`
+      <div class='keypad'>
```
[the remark this answers is not in the document]
Comment (Pingu1337) [reply frontend/components/keypad-dialog.js:38]: nope its on :61

File: frontend/components/keypad-dialog.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+          <button value='8' accessKey='keyBtnUnlock' class='num-button'>8</button>
+          <button value='9' accessKey='keyBtnUnlock' class='num-button'>9</button>
+        </div>
+        <div class='row'>
+          <button accessKey='UnlockBtn' class='lock-button' id='unlock-button'>Unlock</button>
+        </div>
+      </div>
+    `
+    } else {
+      return this.html`
```
[the remark this answers is not in the document]
Comment (Pingu1337) [reply frontend/components/keypad-dialog.js:64]: This was my initial thought, i hardcoded it to be able to test the unlock feature. Will be looking into making a better looking solution with less code.

File: frontend/components/message-preview.js [hunk 3]
```diff
@@ -21,14 +27,42 @@ Tonic.add(class MessagePreview extends Tonic {
     return processText(text, rant)
   }
 
-  render () {
+  async sleep (ms) {
```
[the remark this answers is not in the document]
Comment (Pingu1337) [reply frontend/components/message-preview.js:30]: yes, forgot to remove it during refactoring

File: frontend/components/keypad-dialog.js [hunk 4]
```diff
@@ -96,98 +109,52 @@ Tonic.add(class KeypadDialog extends Tonic {
...
-  if (unlock !== '') { return x } else {
-    waitForInput(text)
-  }
+  const state = await next($unlockState)
+  if (state) {
+    return get($text)
+  } else return waitForInput()
 }
 
 export async function promptPIN (text) {
```
Comment (Pingu1337) [inline frontend/components/keypad-dialog.js:148]: An awaitable function that waits for user to input secret PIN before decrypting and returning the message to render.

**Review by telamon — APPROVED** [review]
Nice!
This was a big feature, in retrospect it would've been nice to split the keypad dialog component and blockend encryption into two separate PR's/features.

**Pull request merged**